### PR TITLE
evm: add missing genesis fields and export genesis state logic

### PR DESCRIPTION
**Pull request by fedekunze**
Closes: #XXX

## Description

- Add missing evm `GenesisState` fields
- Export the EVM state for upgrades and hardforks.

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: x/evm/genesis.go [hunk 1]
```diff
@@ -1,20 +1,47 @@
...
 		k.CreateGenesisAccount(ctx, record)
 	}
 	return []abci.ValidatorUpdate{}
 }
 
 // ExportGenesis exports genesis state
-func ExportGenesis(ctx sdk.Context, _ Keeper) GenesisState {
-	return GenesisState{Accounts: nil}
+func ExportGenesis(ctx sdk.Context, k Keeper, ak types.AccountKeeper) GenesisState {
+	var ethGenAccounts []types.GenesisAccount
```
Comment (golangcibot) [inline x/evm/genesis.go]: Consider preallocating `ethGenAccounts` (from `prealloc`)

File: x/evm/keeper/keeper.go [hunk 2]
```diff
@@ -92,274 +91,31 @@ func (k *Keeper) SetBlockBloomMapping(ctx sdk.Context, bloom ethtypes.Bloom, hei
...
+func (k *Keeper) SetTransactionLogs(ctx sdk.Context, hash []byte, logs []*ethtypes.Log) {
 	store := ctx.KVStore(k.blockKey)
-	encLogs, err := types.EncodeLogs(logs)
-	if err != nil {
-		return err
-	}
-	store.Set(types.LogsKey(hash), encLogs)
 
-	return nil
+	bz := k.cdc.MustMarshalBinaryLengthPrefixed(logs)
```
Comment (noot) [inline x/evm/keeper/keeper.go:107]: will this panic if `logs` is nil?

**Review by noot — COMMENTED** [review]
looks good, could you add instructions for testing the changes?

File: x/evm/keeper/keeper.go (hunk 2 above)
Comment (fedekunze) [reply x/evm/keeper/keeper.go:107]: ah yeah I should add that check

**Comment by fedekunze** [comment]
It's the same instructions but with `emintd`

https://hub.cosmos.network/master/gaia-tutorials/upgrade-node.html#exporting-state-to-a-new-genesis-locally

**Comment by noot** [comment]
@fedekunze I tried running the instructions and got an error:
```
$ emintd export --for-zero-height --height=1 > new_genesis.json
panic: app.baseKey expected to be nil; duplicate init?

goroutine 1 [running]:
github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).initFromMainStore(0xc000e49900, 0xc00012dfd0, 0x0, 0x0)
	/home/elizabeth/go/pkg/mod/github.com/cosmos/cosmos-sdk@v0.34.4-0.20191213112149-d7b0f4b9b4fb/baseapp/baseapp.go:313 +0x2a0
github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).LoadVersion(0xc000e49900, 0x1, 0xc00012dfd0, 0x4, 0xc0003862f8)
	/home/elizabeth/go/pkg/mod/github.com/cosmos/cosmos-sdk@v0.34.4-0.20191213112149-d7b0f4b9b4fb/baseapp/baseapp.go:291 +0x80
github.com/cosmos/ethermint/app.(*EthermintApp).LoadHeight(...)
	/home/elizabeth/go/src/github.com/ChainSafe/ethermint/app/ethermint.go:309
main.exportAppStateAndTMValidators(0x180fd40, 0xc000f80a00, 0x1826280, 0xc0005001f8, 0x0, 0x0, 0x1, 0xc000f80901, 0x23bbaf8, 0x0, ...)
	/home/elizabeth/go/src/github.com/ChainSafe/ethermint/cmd/emintd/main.go:105 +0x100
github.com/cosmos/cosmos-sdk/server.ExportCmd.func1(0xc000e7bb80, 0xc000e4db60, 0x0, 0x2, 0x0, 0x0)
	/home/elizabeth/go/pkg/mod/github.com/cosmos/cosmos-sdk@v0.34.4-0.20191213112149-d7b0f4b9b4fb/server/export.go:65 +0x53d
github.com/spf13/cobra.(*Command).execute(0xc000e7bb80, 0xc000e4db20, 0x2, 0x2, 0xc000e7bb80, 0xc000e4db20)
	/home/elizabeth/go/pkg/mod/github.com/spf13/cobra@v0.0.7/command.go:838 +0x460
github.com/spf13/cobra.(*Command).ExecuteC(0xc0000e8840, 0x13ee756, 0xc000ec5e48, 0x5893e2)
	/home/elizabeth/go/pkg/mod/github.com/spf13/cobra@v0.0.7/command.go:943 +0x317
github.com/spf13/cobra.(*Command).Execute(...)
	/home/elizabeth/go/pkg/mod/github.com/spf13/cobra@v0.0.7/command.go:883
github.com/tendermint/tendermint/libs/cli.Executor.Execute(0xc0000e8840, 0x15cdbb8, 0x13d0971, 0x10)
	/home/elizabeth/go/pkg/mod/github.com/tendermint/tendermint@v0.32.8/libs/cli/setup.go:89 +0x3c
main.main()
	/home/elizabeth/go/src/github.com/ChainSafe/ethermint/cmd/emintd/main.go:88 +0x760
```

**Comment by thomasmodeneis** [comment]
It seems like a test for running the cli should be added to the integration tests.

**Review by thomasmodeneis — DISMISSED** [review]
Looks good, since @noot reported the error, I think it make sense to add a test case for running the cmd: `emintd export --for-zero-height --height=1 > new_genesis.json` as part of the CI #268 in order to prevent it to break in the future.

**Comment by fedekunze** [comment]
> I think it make sense to add a test case for running the cmd: emintd export --for-zero-height --height=1 > new_genesis.json as part of the CI #268 in order to prevent it to break in the future.

Simulations handle that test case. The current export error seems to be unrelated to these changes (i.e it's from the SDK). I'd say we merge this PR as is and fix the export logic afterwards since this PR also includes the missing genesis fields for the evm module.

**Review by noot — APPROVED** [review]
I'm still seeing the same error I mentioned before but let's look into that later, everything else looks good
